crates/pathfinder/src/storage/schema/revision_0018.rs

```diff
@@ -0,0 +1,202 @@
+    let fk_config = tx
+        .db_config(SQLITE_DBCONFIG_ENABLE_FKEY)
+        .context("Reading FK configuration")?;
+    tx.set_db_config(SQLITE_DBCONFIG_ENABLE_FKEY, false)
+        .context("Disabling foreign-key enforcement")?;
+
+    migrate_blocks(tx).context("Migrating starknet_blocks table")?;
+    create_canonical_chain(tx).context("Creating canonical_blocks table")?;
+    migrate_events(tx).context("Migrating starknet_events table")?;
+
+    tx.set_db_config(SQLITE_DBCONFIG_ENABLE_FKEY, fk_config)
+        .context("Setting FK enforcement to pre-migration value")?;
+
+    Ok(())
+}
+
+fn migrate_blocks(tx: &Transaction<'_>) -> anyhow::Result<()> {
+    tx.execute(
+        r"-- Stores StarkNet block headers.
+CREATE TABLE starknet_blocks_new (
+    hash      BLOB    PRIMARY KEY NOT NULL,
+    number    INTEGER NOT NULL,
+    root      BLOB    NOT NULL,
+    timestamp INTEGER NOT NULL, 
+    gas_price BLOB    NOT NULL,
+    sequencer_address BLOB NOT NULL,
+    version_id INTEGER REFERENCES starknet_versions(id)
+)",
+        [],
+    )
+    .context("Creating new table")?;
+
+    tx.execute(
+        r"INSERT INTO starknet_blocks_new(hash,number,root,timestamp,gas_price,sequencer_address,version_id) 
+                                   SELECT hash,number,root,timestamp,gas_price,sequencer_address,version_id FROM starknet_blocks",
+        [],
+    )
+    .context("Copying data")?;
+
+    tx.execute("DROP TABLE starknet_blocks", [])
+        .context("Dropping old table")?;
+
+    tx.execute(
+        "ALTER TABLE starknet_blocks_new RENAME TO starknet_blocks",
+        [],
+    )
+    .context("Renaming table")?;
+
+    tx.execute(
+        "CREATE INDEX starknet_blocks_block_number ON starknet_blocks(number)",
+        [],
+    )
+    .context("Creating block_number index")?;
+
+    Ok(())
+}
+
+/// Re-creates the `starknet_events` events table, updating the `block_number` FK
+/// to reference the `canonical_blocks` table instead.
+fn migrate_events(tx: &Transaction<'_>) -> anyhow::Result<()> {
+    let row_count: usize = tx
+        .query_row("SELECT count(1) FROM starknet_events", [], |r| r.get(0))
+        .context("Count rows in starknet_events table")?;
+
+    if row_count > 0 {
+        tracing::info!(
+            %row_count,
+            "Migrating events table, this may take a while",
+        );
+    }
+
+    tx.execute(
+        r"CREATE TABLE starknet_events_new (
+    id INTEGER PRIMARY KEY NOT NULL,
+    block_number  INTEGER NOT NULL,
+    idx INTEGER NOT NULL,
+    transaction_hash BLOB NOT NULL,
+    from_address BLOB NOT NULL,
+    -- Keys are represented as base64 encoded strings separated by space
+    keys TEXT,
+    data BLOB,
+    FOREIGN KEY(block_number) REFERENCES canonical_blocks(number) ON DELETE CASCADE
+)",
+        [],
+    )
+    .context("Creating new table")?;
+
+    tx.execute(
+        r"-- Copy rowids to be sure that starknet_events_keys still references valid rows
```

Comment:
```suggestion
        r"
```
This comment is not needed with explicit `id` which is copied ofc.

Reply:
Removed thanks